### PR TITLE
Refactor Tree interface

### PR DESCRIPTION
**Pull request by itaiad200**
This change is [Reviewable](https://reviewable.io/reviews/treeverse/lakefs/1059)

### CODE REVIEW

File: graveler/committed/tree/tree.go [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+// Tree is a sorted slice of parts with no overlapping between the parts
+// nolint: structcheck, unused
+type Tree struct {
+	ID    graveler.TreeID
+	Parts []Part
+}
+
+// Repo is an abstraction for a repository of trees that exposes operations on them
+type Repo interface {
+	GetTree(treeID graveler.TreeID) (Tree, error)
```
Comment (johnnyaug) [inline graveler/committed/tree/tree.go:24]: If tree is not an interface, better return a pointer. And receive it as a pointer in all other methods

Comment (johnnyaug) [reply graveler/committed/tree/tree.go:24]: same for ValueRecord

File: graveler/committed/tree/tree.go [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+	// (in other words - values must be entered sorted by key order)
+	// if last insertion operation was AddParts - record key must be greater than any key in the added parts.
+	WriteRecord(record graveler.ValueRecord) error
+
+	// AddParts adds complete parts to the tree at the current insertion point.
+	// Added parts must not contain keys smaller than last previously written value.
+	AddParts(parts []Part) error
+
+	// FlushIterToTree writes the content of an iterator to the tree.
+	FlushIterToTree(iter graveler.ValueIterator) error
```
Comment (johnnyaug) [inline graveler/committed/tree/tree.go:63]:
```suggestion
	WriteIterator(iter graveler.ValueIterator) error
```

File: graveler/committed/tree/tree.go [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+
+	// GetValue finds the matching graveler.ValueRecord in the tree with the treeID
+	GetValue(treeID graveler.TreeID, key graveler.Key) (graveler.ValueRecord, error)
+
+	// NewTreeWriter returns a writer that is used for creating new trees
+	NewTreeWriter() Writer
+
+	// NewIteratorFromTreeID accepts a tree ID, and returns an iterator
+	// over the tree from the first value GE than the from
+	NewIteratorFromTreeID(treeID graveler.TreeID, from graveler.Key) (graveler.ValueIterator, error)
```
Comment (johnnyaug) [inline graveler/committed/tree/tree.go:34]:
```suggestion
	NewIterator(treeID graveler.TreeID, from graveler.Key) (graveler.ValueIterator, error)
```

File: graveler/committed/tree/tree.go [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+	// NewIteratorFromTreeParts accept a tree in memory, returns an iterator
+	// over the tree from the first value GE than the from
+	NewIteratorFromTreeObject(tree Tree, from graveler.Key) (graveler.ValueIterator, error)
+
+	// GetIterForPart accepts a tree ID and a reading start point. it returns an iterator
+	// positioned at the start point. When Next() will be called, first value that is GE
+	// than the from key will be returned
+	GetIterForPart(partID sstable.ID, from graveler.Key) (graveler.ValueIterator, error)
```
Comment (johnnyaug) [inline graveler/committed/tree/tree.go:43]:
```suggestion
	NewPartIterator(partID sstable.ID, from graveler.Key) (graveler.ValueIterator, error)
```

File: graveler/committed/tree/tree.go [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+
+	// PrepareTreesForDiff accepts the left and right trees of the diff, and finds the common parts which
+	// exist in both trees.
+	// it returns the left and right trees with common parts filtered.
+	PrepareTreesForDiff(LeftTree, RightTree Tree) (Tree, Tree)
```
Comment (johnnyaug) [inline graveler/committed/tree/tree.go:48]:
```suggestion
	RemoveCommonParts(LeftTree, RightTree Tree) (Tree, Tree)
```

File: graveler/committed/tree/tree.go [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+	// NewTreeWriter returns a writer that is used for creating new trees
+	NewTreeWriter() Writer
+
+	// NewIteratorFromTreeID accepts a tree ID, and returns an iterator
+	// over the tree from the first value GE than the from
+	NewIteratorFromTreeID(treeID graveler.TreeID, from graveler.Key) (graveler.ValueIterator, error)
+
+	// NewIteratorFromTreeParts accept a tree in memory, returns an iterator
+	// over the tree from the first value GE than the from
+	NewIteratorFromTreeObject(tree Tree, from graveler.Key) (graveler.ValueIterator, error)
```
Comment (johnnyaug) [inline graveler/committed/tree/tree.go:38]: New
```suggestion
	NewIteratorFromTree(tree Tree, from graveler.Key) (graveler.ValueIterator, error)
```

File: graveler/committed/tree/tree.go [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+
+	// PrepareTreesForDiff accepts the left and right trees of the diff, and finds the common parts which
+	// exist in both trees.
+	// it returns the left and right trees with common parts filtered.
+	PrepareTreesForDiff(LeftTree, RightTree Tree) (Tree, Tree)
+}
+
+// Writer is an abstraction for creating new trees
+type Writer interface {
+	// WriteRecord adds a record to the tree. The value key must be greater than any other key that was written
```
Comment (johnnyaug) [inline graveler/committed/tree/tree.go:53]:
```suggestion
	// WriteRecord adds a record to the tree. The key must be greater than any other key that was written
```

File: graveler/committed/tree/tree.go [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+	// (in other words - values must be entered sorted by key order)
+	// if last insertion operation was AddParts - record key must be greater than any key in the added parts.
+	WriteRecord(record graveler.ValueRecord) error
```
Comment (johnnyaug) [inline graveler/committed/tree/tree.go:56]:
```suggestion
	// if the most recent insertion was using AddParts, the key must be greater than any key in the added parts.
	WriteRecord(record graveler.ValueRecord) error
```

**Review by itaiad200 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/treeverse/lakefs/1059)* status: 0 of 2 files reviewed, 6 unresolved discussions (waiting on @johnnyaug and @tzahij)

___
*[graveler/committed/tree/tree.go, line 34 at r1](https://reviewable.io/reviews/treeverse/lakefs/1059#-MOWPh2yx6d77T7n9LTV-r1-34:-MOWQvIq5N7zcS3acQLG:b-896fix) ([raw file](https://github.com/treeverse/lakefs/blob/951ccae1d4cb4aa0b13f07a5a341b6849f29034d/graveler/committed/tree/tree.go#L34)):*
<details><summary><i>Previously, johnnyaug wrote…</i></summary><blockquote>

```suggestion
	NewIterator(treeID graveler.TreeID, from graveler.Key) (graveler.ValueIterator, error)
```
</blockquote></details>

Done.
___
*[graveler/committed/tree/tree.go, line 38 at r1](https://reviewable.io/reviews/treeverse/lakefs/1059#-MOWPh2yx6d77T7n9LTV-r1-38:-MOWQv_hE-_6aMMicdFj:b-896fix) ([raw file](https://github.com/treeverse/lakefs/blob/951ccae1d4cb4aa0b13f07a5a341b6849f29034d/graveler/committed/tree/tree.go#L38)):*
<details><summary><i>Previously, johnnyaug wrote…</i></summary><blockquote>

New
```suggestion
	NewIteratorFromTree(tree Tree, from graveler.Key) (graveler.ValueIterator, error)
```
</blockquote></details>

Done.
___
*[graveler/committed/tree/tree.go, line 43 at r1](https://reviewable.io/reviews/treeverse/lakefs/1059#-MOWPh2yx6d77T7n9LTV-r1-43:-MOWQzSK5HOMv7_DXwgl:b-896fix) ([raw file](https://github.com/treeverse/lakefs/blob/951ccae1d4cb4aa0b13f07a5a341b6849f29034d/graveler/committed/tree/tree.go#L43)):*
<details><summary><i>Previously, johnnyaug wrote…</i></summary><blockquote>

```suggestion
	NewPartIterator(partID sstable.ID, from graveler.Key) (graveler.ValueIterator, error)
```
</blockquote></details>

Done.
___
*[graveler/committed/tree/tree.go, line 48 at r1](https://reviewable.io/reviews/treeverse/lakefs/1059#-MOWPh2yx6d77T7n9LTV-r1-48:-MOWR1I84Z-bKnSlTSUB:b-896fix) ([raw file](https://github.com/treeverse/lakefs/blob/951ccae1d4cb4aa0b13f07a5a341b6849f29034d/graveler/committed/tree/tree.go#L48)):*
<details><summary><i>Previously, johnnyaug wrote…</i></summary><blockquote>

```suggestion
	RemoveCommonParts(LeftTree, RightTree Tree) (Tree, Tree)
```
</blockquote></details>

Done.
___
*[graveler/committed/tree/tree.go, line 53 at r1](https://reviewable.io/reviews/treeverse/lakefs/1059#-MOWPh2yx6d77T7n9LTV-r1-53:-MOWRB9aEFwkGE-UCMbU:b-896fix) ([raw file](https://github.com/treeverse/lakefs/blob/951ccae1d4cb4aa0b13f07a5a341b6849f29034d/graveler/committed/tree/tree.go#L53)):*
<details><summary><i>Previously, johnnyaug wrote…</i></summary><blockquote>

```suggestion
	// WriteRecord adds a record to the tree. The key must be greater than any other key that was written
```
</blockquote></details>

Done.
___
*[graveler/committed/tree/tree.go, line 56 at r1](https://reviewable.io/reviews/treeverse/lakefs/1059#-MOWPh2yx6d77T7n9LTV-r1-56:-MOWRJkP7AUiXM15bhg8:b-896fix) ([raw file](https://github.com/treeverse/lakefs/blob/951ccae1d4cb4aa0b13f07a5a341b6849f29034d/graveler/committed/tree/tree.go#L56)):*
<details><summary><i>Previously, johnnyaug wrote…</i></summary><blockquote>

```suggestion
	// if the most recent insertion was using AddParts, the key must be greater than any key in the added parts.
	WriteRecord(record graveler.ValueRecord) error
```
</blockquote></details>

Done.

**Review by ozkatz — APPROVED** [review]
LGTM

**Pull request merged**